core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -83,6 +84,7 @@ class SparkContext(config: SparkConf) extends Logging {
 
   val startTime = System.currentTimeMillis()
 
+  private[spark] val stopping: AtomicBoolean = new AtomicBoolean(false)
```

Comment:
what's the rationale of adding this?

Reply:
This is to prevent multiple concurrent attempts to stop and ensure that the stop (and the hooks) are executed only once (similar to the `stopped` variable). But we cannot use the `stopped` variable because this is used to determine the spark context stopped status. So if thats set when the hooks are run, the query stop fails with exceptions since there are checks to make sure the spark context is alive before the query can stop.

Reply:
Just 2 cents: not sure about preference, but if we mind about having two different fields for dealing with stopping context, we could just define state by int - 0 (no-request) and 1 (stop-in-progress) and 2 (stopped) - and leverage one AtomicInteger.

Reply:
Thanks for the input. I didn't want to get into too much refactoring since I saw that variable being referred from a few places so just used a different one. It agree it would be cleaner to use a single variable to track the state so I made the change.